client/state/plugins/installed/selectors.ts

```diff
@@ -0,0 +1,382 @@
+ */
+export const getAllPluginsIndexedByPluginSlug = createSelector(
+	( state: AppState ) => {
+		if ( isRequestingForAllSites( state ) ) {
+			return {};
+		}
+
+		return getSiteIdsThatHavePlugins( state ).reduce(
+			( plugins: { [ key: string ]: Plugin }, siteId ) => {
+				if ( isRequesting( state, siteId ) ) {
+					return plugins;
+				}
+
+				const installedPlugins = state.plugins.installed.plugins as InstalledPlugins;
+
+				const pluginsForSite = installedPlugins[ siteId ] || [];
+				pluginsForSite.forEach( ( plugin: InstalledPluginData ) => {
+					const sitePluginProperties = [ 'active', 'autoupdate', 'update', 'version' ];
+					const sitePluginInfo = pick( plugin, sitePluginProperties );
+					const otherPluginInfo = omit( plugin, sitePluginProperties );
+
+					plugins[ plugin.slug ] = {
+						...plugins[ plugin.slug ],
+						...otherPluginInfo,
+						sites: {
+							...plugins[ plugin.slug ]?.sites,
+							[ siteId ]: sitePluginInfo,
+						},
+					};
+				} );
+				return plugins;
+			},
+			{}
+		);
+	},
+	( state ) => [
+		isRequestingForAllSites( state ),
+		getSiteIdsThatHavePlugins( state ),
+		...getSiteIdsThatHavePlugins( state ).map( ( siteId ) => isRequesting( state, siteId ) ),
+	]
+);
+
+/**
+ * The plugins here differ from the plugin objects found on state.plugins.installed.plugins in that each plugin
+ * object has data from all the different sites on it, whereas on state.plugins.installed.plugins they only have
+ * the state for the site which index they are under. The objects here are the same as those returned from
+ * getAllPluginsIndexedByPluginSlug, except they are indexed by siteId.
+ */
+export const getAllPluginsIndexedBySiteId = createSelector(
+	( state ) => {
+		const allPluginsIndexedByPluginSlug = getAllPluginsIndexedByPluginSlug( state );
+
+		return Object.values( allPluginsIndexedByPluginSlug ).reduce(
+			(
+				pluginsIndexedBySiteId: { [ siteId: number ]: { [ pluginSlug: string ]: Plugin } },
+				plugin
+			) => {
+				Object.keys( plugin.sites )
+					.map( ( siteId ) => Number( siteId ) )
+					.forEach( ( siteId ) => {
+						pluginsIndexedBySiteId[ siteId ] = {
+							...pluginsIndexedBySiteId[ siteId ],
+							[ plugin.slug ]: plugin,
+						};
+					} );
+
+				return pluginsIndexedBySiteId;
+			},
+			{}
+		);
+	},
+	( state ) => [ getAllPluginsIndexedByPluginSlug( state ), getSiteIdsThatHavePlugins( state ) ]
+);
+
+export const getFilteredAndSortedPlugins = createSelector(
+	( state: AppState, siteIds: number[], pluginFilter?: PluginFilter ) => {
+		const allPluginsIndexedBySiteId = getAllPluginsIndexedBySiteId( state );
+
+		// Properties on the objects in allPluginsIndexedBySiteId will be modified and the
+		// selector memoization always returns the same object, so use cloneDeep to avoid
+		// altering it for everyone.
+		const allPluginsForSites: { [ pluginSlug: string ]: Plugin } = cloneDeep(
```

Comment:
If I recall correctly some tests were failing because they called this selector with different sets of site Ids, and the mutation done in this selector was changing the objects returned from `getAllPluginsIndexedBySiteId` depending on the site Ids passed. The `cloneDeep` fixed that.

The object has to be mutated because of the behavior under the "Filter the sites object on the plugins so that only data for the requested siteIds is present" comment, which is assumed by a lot of component code. I think ideally that wouldn't be an assumption of the component code because the plugin should still have all its sites regardless of which sites were requested, but I looked into it and this assumption would be very difficult to change. 

The memoization is still useful since both selectors are memoized. `allPluginsIndexedBySiteId` will only be calculated once per fetch of plugins, and `getFilteredAndSortedPlugins` will be calculated once for each set of siteIds and filter passed to it, and only the data for the sites that are requested is getting cloned.

Reply:
I see, thanks for clarifying.

Perhaps we could use [`structuredClone()`](https://developer.mozilla.org/en-US/docs/Web/API/structuredClone) instead of introducing yet another Lodash usage?

Reply:
`structuredClone` is added in node 17, but Calypso still supports 16 according to package.json and the tests seem to be running on 16. So if I switch to `structuredClone` it will crash in some supported environments.

Reply:
We could add a polyfill for the server. There are a few modules that can help with that (https://www.npmjs.com/package/realistic-structured-clone). It's definitely worth the effort, given that we're getting rid of Lodash and we do our best not to introduce additional usages. It's also worth double-checking if on the server we really need this module's functionality, or it's necessary just to compile and will never execute on the server. If we have the latter case, we could just add a workaround to noop that module for the server.

Reply:
We just updated to Node 18 last week -- the main Calypso server runs Node 18 now, and I don't have any plans to maintain backwards compatibility with Node 16

Reply:
I did some more investigation, and it seems now the issue is in `jest`. Jest 27 [had a bug](https://github.com/facebook/jest/issues/12628) where `structuredClone` wasn't available and it is fixed in Jest 28.

Reply:
This might be a good opportunity to upgrade Jest - this is something @jsnajdr has been looking to do in Gutenberg as well. In the meantime, if tests are the only reason for not using `structuredClone` we could temporarily do something like this in the test:

```